In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import accuracy_score  
from sklearn.metrics import recall_score  
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.linear_model import LogisticRegression 
from sklearn import neighbors
from sklearn import tree
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB
import matplotlib.pyplot as plt

In [2]:
def get_the_predicted_bugs(name, num1, num2):
    All_Data = pd.read_csv(name, usecols = [i for i in range(num1, num2)]).values
    return All_Data

In [3]:
data = get_the_predicted_bugs('modified_all_data.csv', 2, 45)

## Decision Trees

## best accuracy_score

In [4]:
data_random = np.random.permutation(data)
x = data_random[:, [i for i in range(data.shape[1] - 1)]]
y = data_random[:, -1]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
dt_clf = tree.DecisionTreeClassifier()

In [5]:
entropy_thresholds = np.linspace(0, 1, 10)
gini_thresholds = np.linspace(0, 0.2, 10)

param_grid = [
              {'criterion': ['entropy', 'gini'],
               'min_impurity_decrease': entropy_thresholds,
               'max_depth': np.arange(2,10),
               'min_samples_split': np.arange(2,30,2)},
              {'criterion': ['gini'],
               'min_impurity_decrease': gini_thresholds,
               'max_depth': np.arange(2,10),
               'min_samples_split': np.arange(2,30,2)}
             ]

grid_search = GridSearchCV(dt_clf, param_grid, scoring = 'accuracy', n_jobs = -1, cv = 5)

In [6]:
grid_search.fit(x_train, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best'),
       fit_params=None, iid=True, n_jobs=-1,
       param_grid=[{'criterion': ['entropy', 'gini'], 'min_impurity_decrease': array([0.     , 0.11111, 0.22222, 0.33333, 0.44444, 0.55556, 0.66667,
       0.77778, 0.88889, 1.     ]), 'max_depth': array([2, 3, 4, 5, 6, 7, 8, 9]), 'min_samples_split': array([ 2,  4,  6,  8, 10, 12, 14, 16, 18, 20, 22, 24, ...6, 7, 8, 9]), 'min_samples_split': array([ 2,  4,  6,  8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28])}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='accuracy', verbose=0)

In [7]:
grid_search.best_params_

{'criterion': 'entropy',
 'max_depth': 2,
 'min_impurity_decrease': 0.1111111111111111,
 'min_samples_split': 2}

In [8]:
y_pre = grid_search.best_estimator_.predict(x_test)
accuracy_score(y_test, y_pre), recall_score(y_test, y_pre)

(0.8664440734557596, 0.0)

In [10]:
y_pre

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0.

By Decision Trees, aftering tuning the parameters for getting the best accuracy_score, all the predicted results are equal to 0.

## best recall_score

In [21]:
data_random = np.random.permutation(data)
x = data_random[:, [i for i in range(data.shape[1] - 1)]]
y = data_random[:, -1]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
dt_clf = tree.DecisionTreeClassifier()

In [22]:
entropy_thresholds = np.linspace(0, 1, 10)
gini_thresholds = np.linspace(0, 0.2, 10)

param_grid = [
              {'criterion': ['entropy', 'gini'],
               'min_impurity_decrease': entropy_thresholds,
               'max_depth': np.arange(2, 100, 10),
               'min_samples_split': np.arange(2,30,2)},
              {'criterion': ['gini'],
               'min_impurity_decrease': gini_thresholds,
               'max_depth': np.arange(2, 100, 10),
               'min_samples_split': np.arange(2,30,2)}
             ]

grid_search = GridSearchCV(dt_clf, param_grid, scoring = 'recall', n_jobs = -1, cv = 5)

In [23]:
grid_search.fit(x_train, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best'),
       fit_params=None, iid=True, n_jobs=-1,
       param_grid=[{'criterion': ['entropy', 'gini'], 'min_impurity_decrease': array([0.     , 0.11111, 0.22222, 0.33333, 0.44444, 0.55556, 0.66667,
       0.77778, 0.88889, 1.     ]), 'max_depth': array([ 2, 12, 22, 32, 42, 52, 62, 72, 82, 92]), 'min_samples_split': array([ 2,  4,  6,  8, 10, 12, 14, 16, ...72, 82, 92]), 'min_samples_split': array([ 2,  4,  6,  8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28])}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='recall', verbose=0)

In [24]:
grid_search.best_params_

{'criterion': 'gini',
 'max_depth': 72,
 'min_impurity_decrease': 0.0,
 'min_samples_split': 2}

In [25]:
y_pre = grid_search.best_estimator_.predict(x_test)
accuracy_score(y_test, y_pre), recall_score(y_test, y_pre)

(0.8063439065108514, 0.2916666666666667)

## get the average accuracy_score

In [26]:
accuracy_score_list = []
recall_score_list = []
for i in range(100):
    data_random = np.random.permutation(data)
    x = data_random[:, [i for i in range(data.shape[1] - 1)]]
    y = data_random[:, -1]
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
    dt_clf = tree.DecisionTreeClassifier(criterion = 'gini', max_depth = 72, min_impurity_decrease = 0, min_samples_split = 2)
    dt_clf.fit(x_train, y_train)
    y_pre = dt_clf.predict(x_test)
    accuracy_score_list.append(accuracy_score(y_test, y_pre))
    recall_score_list.append(recall_score(y_test, y_pre))

In [27]:
np.mean(accuracy_score_list)

0.7922871452420703

In [28]:
np.mean(recall_score_list)

0.2932990520826039

By  Decision Trees, aftering tuning the parameters for getting the best recall_score, the average accuracy_score of 100 times is about is about 0.79 and the average recall_score of 100 times is about 0.29.